src/main/kotlin/io/kotlintest/SpecExecutor.kt

```diff
@@ -0,0 +1,80 @@
+package io.kotlintest
+
+import org.junit.platform.engine.EngineExecutionListener
+import org.junit.platform.engine.TestDescriptor
+import org.junit.platform.engine.TestExecutionResult
+
+interface SpecExecutor {
+  fun execute(specDescriptor: SpecDescriptor, listener: EngineExecutionListener)
+}
+
+// a basic implementation that supports spec interceptors, traverses the specs, and for each detected test invokes 'run'
+abstract class AbstractSpecExecutor : SpecExecutor {
+
+  protected abstract fun run(testCase: TestCaseDescriptor, listener: EngineExecutionListener)
+
+  // the initial interceptor just invokes the intercept method in the spec itself, which in
+  // turn invokes the chain.
+  private val initialInterceptor = { spec: Spec, chain: () -> Unit ->
+    spec.interceptSpec({ chain() })
+  }
+
+  private fun interceptorChain(spec: Spec) = createInterceptorChain(spec.specInterceptors, initialInterceptor)
+
+  private fun <CONTEXT> createInterceptorChain(
+      interceptors: Iterable<(CONTEXT, () -> Unit) -> Unit>,
+      initialInterceptor: (CONTEXT, () -> Unit) -> Unit): (CONTEXT, () -> Unit) -> Unit {
+    return interceptors.reversed().fold(initialInterceptor) { a, b ->
+      { context: CONTEXT, testCase: () -> Unit ->
+        b(context, { a.invoke(context, { testCase() }) })
+      }
+    }
+  }
+
+  private fun run(descriptor: TestDescriptor, listener: EngineExecutionListener) {
```

Comment:
`run` and `execute` sound very similar without explaining the difference. More describing names would be helpful.